doc/how-to-freeze-versions.rst

```diff
@@ -0,0 +1,247 @@
+unexpectedly change.  You can still update your dependencies, but you have to do
+it on purpose, by modifying or by deleting and regenerating
+``cabal.project.freeze`` file, and in the meantime you are guaranteed no
+surprises will happen.
+
+This consistency can be valuable as it ensures that all teammates, deployments,
+and continuous integration are installing the exactly same dependencies.  So if
+you are running and testing the code on your local machine, you are guaranteed
+that your teammate and your continuous integration will be running the exact
+same code, and that at the end that exact same code will get deployed.
+
+A ``.freeze`` file can be good to have when developing for yourself or within a
+private team.  If anyone using it can somehow have different inputs to
+the solver then the ``.freeze`` file can be troublesome.  It can prevent the
+solver from finding a different version of a dependency that would satisfy a
+different architecture or a different compiler version and boot libraries.
+
+.. Warning::
+
+    If publishing a package to Hackage, not matter what kind of component it
+    contains, don't include a ``.freeze`` file, don't add it to any field of the
+    package description that would have ``cabal sdist`` include it in the
+    ``.tar.gz``. In general, don't include anything in the package description
+    that relates to the project environment, like ``cabal.project`` or
+    ``cabal.project.local``.
+
+Freezing workflows
+^^^^^^^^^^^^^^^^^^
+
+.. Warning::
+    For each of these workflows, you may have to first delete the
+    ``index-state`` line from ``cabal.project`` (and from
+    ``cabal.project.freeze`` if it exists) and then run ``cabal update`` to
+    ensure that cabal will have newer versions to re-resolve the dependencies
+    with. Alternatively, you can run ``cabal update
+    --ignore-project``.
+
+Freeze
+    If the ``cabal.project.freeze`` file doesn't exist, generating one is a
+    great way to see what versions of dependencies are currently being used even
+    if you choose to discard the ``.freeze`` file after inspecting it.
+
+Thaw, Freeze
+    If you changed the version ranges of any of the dependencies in any of your
+    project's package descriptions, in any ``.cabal`` file, then delete the
+    ``cabal.project.freeze`` file if it already exists and run ``cabal freeze``
+    to generate fresh version of ``cabal.project.freeze``.
+
+Freeze, Freeze (Freezing Harder)
+    You might in some cases want to skip deletion of ``cabal.project.freeze``,
+    but keep in mind that in that case ``cabal freeze`` will use existing
+    ``cabal.project.freeze`` when resolving dependencies, therefore not updating
+    any existing dependencies, only adding new ones.
+
+Partial Thaw, Freeze
+    If you want to a pick up a different version of a single dependency, you can
+    delete its constraint from ``cabal.project.freeze`` and then run ``cabal
+    freeze`` again.
```

Comment:
I would suggest organizing these by the goals that the user is trying to solve, e.g.,

> Check which versions of dependencies Cabal has selected
>     If the ``cabal.project.freeze`` file doesn't exist, ...
>
> Update the versions of dependencies listed in the ``cabal.project.freeze`` file
>     To bump the versions of dependencies listed in the freeze file (for example, because you changed the version range in a ``.cabal`` file), first delete ...
>
> Add a new dependency to the ``cabal.project.freeze`` file
>     If you have added a new dependency to a ``.cabal`` file, simply re-run ...
>
> Updating a single dependency
>     ...

Reply:
Thanks @langston-barrett for this suggestion. I've redone this section on worflows.

Reply:
Thanks, the new version is nice!